com.playeveryware.eos/Runtime/Core/Common/SetOfNamed.cs

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2024 PlayEveryWare
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+namespace PlayEveryWare.Common
+{
+    using System.Collections.Generic;
+    using System;
+
+    /// <summary>
+    /// Class that stores a unique set of values, where each value has a name
+    /// associated to it. Items are sorted by the name. No two items of the same
+    /// value can be added (regardless of if they have different names). Also,
+    /// no two items can have the same name (regardless of if they have
+    /// different values).
+    /// </summary>
+    /// <typeparam name="T">The type being wrapped with a name.</typeparam>
+    public class SetOfNamed<T> : HashSet<Named<T>> where T : IEquatable<T>
```

Comment:
I believe HashSet uses IEqualityComparer<T> if it is available. If it isn't available, then it's doing a more expensive check *I think*. Can you consider requiring T to implement IEqualityComparer?

If you'd like I can whip up a demo that shows the utility of this suggestion.

Reply:
Per offline conversation - given that these collections will 99.99% of the time contain at most 2 items, the cost of the equality check is not a significant enough concern to warrant further development. But it's good to be aware of, thank you for calling it to our attention! :grin:

Reply:
Sure, I can agree with that.

I *want* to have the time to make everything as optimally implemented as possible. There's this story in my head where a user who is relatively new to programming is learning through struggle, continuously running into walls until they can accomplish what they want. For this theoretical user, I'd want every bit of code they see to be a learning moment. If they want to implement a HashSet, similar to how you've done here, they'd learn how to do it *the right way* by looking at it.

But sometimes *the right way* is to make something that fits the need, not something that fits every theoretical angle.